### ISSUE TITLE
Consecutive file uploads using SendKeys with Chrome v75 or later. (C#)

### ISSUE THREAD

**Issue by werfmark** [SeleniumHQ/selenium#7472]
## 🐛 Bug Report

Consecutive file uploads to an Input[type=file] element with SendKeys duplicate files uploaded before in Chrome v75 or later. In Chrome v74 this didn't happen.
Specifically uploaded a file will make all files be uploaded again. So a 4th file upload will unintentionally upload file 1, 2 and 3 again. 

## To Reproduce

Detailed steps to reproduce the behavior:

1. Search an "input[type=file]" WebElement
2. Use ChromeWebDriver with Chrome 76.0.3809.87 installed and Selenium.Chrome.WebDriver 76.0.0 and Selenium.WebDriver 3.141.0
3. Use SendKeys twice on the input element to send a file (NOTE: not two files at once but consecutively)

## Expected behavior

The files are uploaded once each.

## Actual behavior

The first file is uploaded twice, the second file once. 

## Test Code C#

```
using Microsoft.VisualStudio.TestTools.UnitTesting;
using OpenQA.Selenium;
using OpenQA.Selenium.Chrome;
using System;
using System.IO;
using System.Reflection;

namespace UnitTestProject1
{
    [TestClass]
    public class UnitTest1
    {
        [TestMethod]
        public void TestMethod1()
        {
            var driver = new ChromeDriver(Directory.GetCurrentDirectory(), new ChromeOptions());

            driver.Manage().Timeouts().ImplicitWait = TimeSpan.FromSeconds(15);

            driver.Url = "https://gofile.io/?t=uploadFiles";

            var path = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
            var pdfFile1 = $"{path}\\Files\\File1.pdf";
            var pdfFile2 = $"{path}\\Files\\File2.pdf";

            var input = driver.FindElementByXPath("//input[@type='file']");

            input.SendKeys(pdfFile1);
            input.SendKeys(pdfFile2);

            Assert.AreEqual(2, driver.FindElementsByXPath("//tbody//tr[@role='row']").Count);
        }
    }
}

```
The two files are random PDFs in a folder called Files in the solution with "Copy always to Output Directory"

##Environment
OS: Windows 7 64bits
Browser: Chrome 76.0.3809.87
Selenium Webdriver: 3.141.0

## Potential Workaround
There is a potential workaround for not having to upload files consecutively by uploading them all at once by using SendKeys with one string with multiple filenames separated by "\n" but this does not work if you're using RemoteWebDriver with LocalFileDetector, see:
https://github.com/SeleniumHQ/selenium/issues/7408

**Comment by jimevans** [comment]
@werfmark I refer you to [my comment](https://github.com/SeleniumHQ/selenium/issues/7489#issuecomment-520882291) on #7489. There is likely little that can be done by the language bindings to change this behavior, unless the `LocalFileDetector` is not correctly handling file lists with embedded `\n` delimiters. I can see a case in the .NET bindings, at least, where that might not be the case.

**Comment by JohnChen0** [comment]
The behavior of ChromeDriver 75 and later has been updated to conform to the latest W3C WebDriver standard. For this particular case, the standard (https://www.w3.org/TR/webdriver/#element-send-keys) specifies that "If multiple is true files are be appended to element’s selected files." So I believe ChromeDriver working correctly by keeping the existing selected files and adding the new ones.

**Comment by werfmark** [comment]
Hmm I don't quite see how
"If multiple is true files are (to) be appended to element's selected files." in the W3C standard means that uploading a second file results in the first file uploaded again. As I understand it uploading a second file should append it to the element's selected files and thus only 2 files are shown uploaded in total? 

I don't see how this W3C webdriver standard wording means this functionality is as intended. Also if it does somehow mean that (I find it hard to grok the exact meaning) I find it an unlogical choice. If you want to add multiple files to an input element consecutively for some reason this means there is no way to do that elegantly now? 

Basically I don't understand this behaviour to be intentional, if it is however it makes no sense?

**Comment by twalpole** [comment]
@werfmark The spec says nothing about files actually being uploaded, just about adding the files to the file input.  As you've noticed the issue is that if you have a change event handler on the file input that automatically uploads attached files it creates a condition where files will get repeatedly uploaded. This because on the first change event there is one file (which gets uploaded but not cleared), then on the second change event two files are attached to the file input so two files get uploaded, etc.  The appending behavior has been argued as making no sense (because it's not something a user can actually do) by a number of people, but it was originally implemented/specified like that for legacy reasons and has stayed in the spec.  

As you've mentioned, uploading multiple files at once is possible, in the spec (\n separated names), and should work even with a File detector (although you may need to execute JS to set the elements `value` property to null before attaching multiples if the input already has files attached).  This does appear to work correctly with the Ruby bindings locally and remotely to the current standalone server release version for both Chrome and Firefox, so it does point towards it possibly being an issue in the bindings you're using.

**Comment by barancev** [comment]
It seems that consecutive file uploads work in as specified.

So I want to reframe this issue: attaching and uploading multiple files at once should be supported in C#, it works locally but not with RemoteWebDriver yet.

**Comment by markhm** [comment]
I've just observed this behaviour described in this issue on Java as well. 

    Selenium Java: org.seleniumhq.selenium/selenium-java:3.141.59
    Chromedriver v79.0.3945.36
    macOS 10.15.2 (19C57) / OpenJDK 13.0.1

The workaround works.

Is there any other place this should be reported...?

**Comment by dethu** [comment]
I also noticed behavior described here using Python.
```
Selenium Python: 3.141.0
Chromedriver v79.0.3945.36
OS: Host machine: Ubuntu 19.04.
Remote WebDriver machine: Ubuntu 18.04.1 LTS (GNU/Linux 4.15.0-74-generic x86_64) with aerokube/selenoid:1.10.0
```
The potential workaround does not work

> There is a potential workaround for not having to upload files consecutively by uploading them all at once by using SendKeys with one string with multiple filenames separated by "\n" but this does not work if you're using RemoteWebDriver with LocalFileDetector, see:
#7408

There is a fix for Java as i found out 
https://github.com/SeleniumHQ/selenium/issues/7408#issuecomment-555341644

 so i have a same question:
> Is there any other place this should be reported...?

**Comment by odonovan-johnd** [comment]
I've just hit this issue on an Angular project, luckily I found this. Unfortunately, I'm not able to do the  "\n" with the way our file uploads are managed. Has there been any work on correcting this behaviour?

**Comment by nandesu** [comment]
> I also noticed behavior described here using Python.
> 
> ```
> Selenium Python: 3.141.0
> Chromedriver v79.0.3945.36
> OS: Host machine: Ubuntu 19.04.
> Remote WebDriver machine: Ubuntu 18.04.1 LTS (GNU/Linux 4.15.0-74-generic x86_64) with aerokube/selenoid:1.10.0
> ```
> 
> The potential workaround does not work
> 
> > There is a potential workaround for not having to upload files consecutively by uploading them all at once by using SendKeys with one string with multiple filenames separated by "\n" but this does not work if you're using RemoteWebDriver with LocalFileDetector, see:
> > #7408
> 
> There is a fix for Java as i found out
> [#7408 (comment)](https://github.com/SeleniumHQ/selenium/issues/7408#issuecomment-555341644)
> 
> so i have a same question:
> 
> > Is there any other place this should be reported...?

Any update on this for Python? \n isn't working for me either.
